modules/platforms/dotnet/Apache.Ignite/Internal/Proto/BinaryTuple/BinaryTupleBuilder.cs

```diff
@@ -0,0 +1,407 @@
+    using System.Text;
+    using Buffers;
+
+    /// <summary>
+    /// Binary tuple builder.
+    /// </summary>
+    internal sealed class BinaryTupleBuilder : IDisposable // TODO: Support all types (IGNITE-15431).
+    {
+        /** Number of elements in the tuple. */
+        private readonly int _numElements;
+
+        /** Size of an offset table entry. */
+        private readonly int _entrySize;
+
+        /** Position of the varlen offset table. */
+        private readonly int _entryBase;
+
+        /** Starting position of variable-length values. */
+        private readonly int _valueBase;
+
+        /** Buffer for tuple content. */
+        private readonly PooledArrayBufferWriter _buffer = new();
+
+        /** Flag indicating if any NULL values were really put here. */
+        private bool _hasNullValues;
+
+        /** Current element. */
+        private int _elementIndex;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BinaryTupleBuilder"/> class.
+        /// </summary>
+        /// <param name="numElements">Capacity.</param>
+        /// <param name="allowNulls">Whether nulls are allowed.</param>
+        /// <param name="totalValueSize">Total value size, -1 when unknown.</param>
+        public BinaryTupleBuilder(int numElements, bool allowNulls = true, int totalValueSize = -1)
+        {
+            Debug.Assert(numElements > 0, "numElements > 0");
+
+            _numElements = numElements;
+
+            int baseOffset = BinaryTupleCommon.HeaderSize;
+            if (allowNulls)
+            {
+                baseOffset += BinaryTupleCommon.NullMapSize(numElements);
+            }
+
+            _entryBase = baseOffset;
+
+            _entrySize = totalValueSize < 0
+                ? 4
+                : BinaryTupleCommon.FlagsToEntrySize(BinaryTupleCommon.ValueSizeToFlags(totalValueSize));
+
+            _valueBase = baseOffset + _entrySize * numElements;
+
+            _buffer.GetSpan(_valueBase);
+            _buffer.Advance(_valueBase);
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether null map is present.
+        /// </summary>
+        public bool HasNullMap => _entryBase > BinaryTupleCommon.HeaderSize;
+
+        /// <summary>
+        /// Gets the current element index.
+        /// </summary>
+        public int ElementIndex => _elementIndex;
+
+        /// <summary>
+        /// Appends a null value.
+        /// </summary>
+        public void AppendNull()
+        {
+            if (!HasNullMap)
+            {
+                throw new InvalidOperationException("Appending a NULL value in binary tuple builder with disabled NULLs");
+            }
+
+            _hasNullValues = true;
+
+            int nullIndex = BinaryTupleCommon.NullOffset(_elementIndex);
+            byte nullMask = BinaryTupleCommon.NullMask(_elementIndex);
+
+            var span = _buffer.GetSpan(nullIndex, 1);
+
+            span[0] |= nullMask;
```

Comment:
Why not just 
```suggestion
            _buffer[nullIndex] |= nullMask;
```

Reply:
`_buffer` is not an array, so we can't do that.